New Python Solutions (#1682)
cpinitiative/usaco-guide: solutions/usaco-832.mdx

---
id: usaco-832
source: USACO Bronze 2018 Open
title: Milking Order
author: Ananth Kashyap
---

## Solution 1

[Official Analysis](http://www.usaco.org/current/data/sol_milkorder_bronze_open18.html)

## Solution 2

> In this problem, there are three cases that we have to consider. The first of these cases is that cow one is fixed. In this case, the solution is simple, as we just output its fixed position. The second case we must consider is if cow one is in the given hierarchy of cows. In this case, we want to place every cow in the hierarchy as early as possible. The last case is when cow one is neither fixed nor in the hierarchy. In this case, we would want to place every cow in the hierarchy as late as possible, allowing for cow one to get the earliest position possible.

Reviewer: It'd be nice to break this down into bullet points.

<LanguageSection>
<PySection>

> 

Reviewer 2: ```suggestion

## Python Implementation

```

> ```py

Reviewer 3: use tabs instead of spaces to indent

file_in = open('milkorder.in')
data = file_in.read().strip().split('\n')
n, m, k = map(int, data[0].split(' '))
hierarchy = list(map(int, data[1].split(' ')))

> fixed = [list(map(int, data[i+2].split(' '))) for i in range(k)]

Reviewer 3: ```suggestion
fixed = [list(map(int, data[i + 2].split(' '))) for i in range(k)]
```

---
id: usaco-832
source: USACO Bronze 2018 Open
title: Milking Order
author: Ananth Kashyap
---

## Solution 1

[Official Analysis](http://www.usaco.org/current/data/sol_milkorder_bronze_open18.html)

## Solution 2

In this problem, there are three cases that we have to consider: 

> - The first of these cases is that cow one is fixed. In this case, the solution is simple, as we just output its fixed position. 

Reviewer 4: ```suggestion
- The first of these cases is when cow one is fixed. In this case, the solution is simple, as we just output its fixed position. 
```

> - The second case we must consider is if cow one is in the given hierarchy of cows. In this case, we want to place every cow in the hierarchy as early as possible. 

Reviewer 4: ```suggestion
- In the second case, we must consider is if cow one is in the given hierarchy of cows. In this case, we want to place every cow in the hierarchy as early as possible. 
```

- The last case is when cow one is neither fixed nor in the hierarchy. In this case, we would want to place every cow in the hierarchy as late as possible, allowing for cow one to get the earliest position possible.
<LanguageSection>
<PySection>

## Python Implementation

```py
file_in = open('milkorder.in')
data = file_in.read().strip().split('\n')
n, m, k = map(int, data[0].split(' '))
hierarchy = list(map(int, data[1].split(' ')))
fixed = [list(map(int, data[i + 2].split(' '))) for i in range(k)]

> cows = [0 for i in range(n)]

Reviewer 4: `cows = [0] * n`

for i in range(k):
  	cows[fixed[i][1] - 1] = fixed[i][0]

> if 1 in cows:

Reviewer 4: can you add comments here? like `# case 1/2/3`

Author: Will do

---
id: usaco-832
source: USACO Bronze 2018 Open
title: Milking Order
author: Ananth Kashyap
---

## Solution 1

[Official Analysis](http://www.usaco.org/current/data/sol_milkorder_bronze_open18.html)

## Solution 2

> In this problem, there are three cases that we have to consider: 

Reviewer 5: maybe do it like this
```
there are 3 cases blah blah blah
1. Cow one is fixed...
2. Cow one is in the govern hierarchy of cows...
3. Cow one is neither fixed nor in the hiearchy
```

Reviewer: ^ I agree

> - The first of these cases is when cow one is fixed. In this case, the solution is simple, as we just output its fixed position. 

Reviewer 6: ```suggestion
- When the first cow is fixed, we output its position. 
```

Reviewer 5: "When cow $1$" would be better imo

Reviewer: ^
Or even "Cow $1$ is fixed: blah blah blah"

Reviewer: Also possibly have ordered list?
```
1. blah blah blah
2. ...
3. ...
```

Unordered is ok tho.

> - In the second case, we must consider is if cow one is in the given hierarchy of cows. In this case, we want to place every cow in the hierarchy as early as possible. 

Reviewer 6: ```suggestion
- When cow 1 is in the given hierarchy of the cows, we should place every cow in the hierarchy as early as possible.
```

Reviewer 6: I think you should elaborate on this.

> - The last case is when cow one is neither fixed nor in the hierarchy. In this case, we would want to place every cow in the hierarchy as late as possible, allowing for cow one to get the earliest position possible.

Reviewer 6: ```suggestion
- When cow 1 is neither fixed nor in the hierarchy, we want to place every cow in the hierarchy as late as possible such that cow 1 gets the earliest possible position.
```

**Time Complexity:** $\mathcal{O}(N(M + N))$

<LanguageSection>
<PySection>

> ## Python Implementation

Reviewer 3: move this outside languagesection (and above the time complexity) and just put ## Implementation

@HarryWangATX i think you suggested this? the template on the guide just has ## Implementation, so i think we should stick with that
i think the reason is that the header should be above time complexity

Author: What should I do

Reviewer 4: > move this outside languagesection (and above the time complexity) and just put ## Implementation

Reviewer: ```
## Implementation

**Time Complexity:** $\mathcal{O}(N(M + N))$

<LanguageSection>

<PySection>

... your code, etc ...
```
